url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+                        pingTimers.get(node_id).cancel();
+                        pingTimers.remove(node_id);
+                    }
+            	}
+
+            	break;
+            case ACK_INDEX:
+            	// the ack_index is different from the ack in that the destination node is not identified by ip but with id.
+            	word = (String)jsonObj.get("keyword");
+                if(indexTimers.containsKey(word)){
+                    indexTimers.get(word).cancel();
+                    indexTimers.remove(word);
+                }
+            	break;
+            case PING:
+            	routingTable.put(sender_id, new Contact(sender_id,ip_address));
+            	singleUpdateRouteTable(sender_id,ip_address);
+            	if(target_id == nodeID){// if current is the suspected dead
+                	message = new JSONObject();
+                	message.put("type", "ACK");
+                	message.put("node_id", String.valueOf(nodeID));
+                	message.put("ip_address",NetworkProperties.nAddress);
+                	address = new InetSocketAddress(packet.getAddress(),packet.getPort());           				
+            		sendQueue.put(jsonToDatagram(message,address)); 
+            	}else{ // else forward
+            		message = jsonObj;
+            		message.put("ip_address",NetworkProperties.nAddress);//current's address changes each hop
+            		if(routingTable.containsKey(target_id)){
+            			address = new InetSocketAddress(routingTable.get(target_id).getIp(),NetworkProperties.nPort);
+            			sendQueue.put(jsonToDatagram(message,address));
+            		}else{
+            			
+            		}             	           				           		 
+            	}           	
+                break;
+                //internal housekeeping. send out pings to stale contacts
+            case REFRESH_ROUTING_TABLE:
+            	//System.out.println("will handle update routing table message");
+               	for (Map.Entry<Integer, Contact> entry : routingTable.entrySet()){
+               		final Contact contact = entry.getValue();
+               		long currTime = System.currentTimeMillis();
+               		if(currTime - contact.getTimestamp() > NetworkProperties.staleTime*1000){	               		
+	        			message = getPingJson(contact);
+	        			address = new InetSocketAddress(contact.getIp(),NetworkProperties.nPort);           				
+				        sendQueue.put(jsonToDatagram(message,address));
+				        Timer pingTimer = new Timer();
+				        pingTimer.schedule(new TimerTask() {
+							//if timer not cancelled before time expires then contact removed
+							public void run() {
+								routingTable.remove(contact.getId());							
+							}
+						}, NetworkProperties.pingWait*1000);
+				        pingTimers.put(contact.getId(), pingTimer);
+	        		}
+               	}
+            	break;
+        }
+        }catch(InterruptedException e){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)